driver-core/src/test/java/com/datastax/driver/core/CCMTestsSupport.java

```diff
@@ -0,0 +1,898 @@
+            this.numberOfNodes = numberOfNodes;
+        }
+
+        private CCMTestContext createAndStartTestContext() {
+            CCMTestContext ccmTestContext = new CCMTestContext(ccmBridgeBuilder, numberOfNodes);
+            ccmTestContext.start();
+            return ccmTestContext;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            CCMTestContextKey that = (CCMTestContextKey) o;
+            return ccmBridgeBuilder.equals(that.ccmBridgeBuilder) && Arrays.equals(numberOfNodes, that.numberOfNodes);
+
+        }
+
+        @Override
+        public int hashCode() {
+            int result = ccmBridgeBuilder.hashCode();
+            result = 31 * result + Arrays.hashCode(numberOfNodes);
+            return result;
+        }
+    }
+
+    private static class CCMTestContext {
+
+        private static final AtomicInteger COUNTER = new AtomicInteger(1);
+
+        private static final AtomicInteger NEXT_PORT = new AtomicInteger(5000);
+
+        private static final ReentrantLock PORT_LOCK = new ReentrantLock();
+
+        private final int id;
+
+        private final CCMBridge.Builder ccmBridgeBuilder;
+
+        private final int[] numberOfNodes;
+
+        private CCMBridge ccmBridge;
+
+        private List<CCMNode> nodes;
+
+        private boolean keepLogs = false;
+
+        private boolean stopped = false;
+
+        public CCMTestContext(CCMBridge.Builder ccmBridgeBuilder, int... numberOfNodes) {
+            this.id = COUNTER.getAndIncrement();
+            this.ccmBridgeBuilder = ccmBridgeBuilder;
+            this.numberOfNodes = numberOfNodes;
+            this.keepLogs = false;
+        }
+
+        private void start() {
+            LOGGER.debug("Starting " + this);
+            // force no start and no nodes
+            ccmBridge = ccmBridgeBuilder
+                    .withoutNodes()
+                    .notStarted()
+                    .build();
+            Runtime.getRuntime().addShutdownHook(new Thread() {
+                @Override
+                public void run() {
+                    CCMTestContext.this.stop();
+                }
+            });
+            Properties props = new Properties();
+            nodes = new ArrayList<CCMNode>();
+            boolean multiDc = numberOfNodes.length > 1;
+            int total = 1;
+            for (int nodesPerDc : numberOfNodes) {
+                total *= nodesPerDc;
+            }
+            PORT_LOCK.lock();
+            try {
+                // storage port must be the same for all nodes in CCM cluster;
+                // find first available for first ip, and hope it will work for all ips
+                int storagePort = nextPort(1);
+                List<TokenRange> ranges = new TokenRange(FACTORY.minToken(), FACTORY.minToken(), FACTORY).splitEvenly(total);
+                int n = 1;
```

Comment:
Ha :) in the meanwhile I looked at CCM code and did the same 100-offset-shift magic – in the end, we have roughly the same code. 

But I confess still don't get why we need this offset; the [docs](https://docs.datastax.com/en/cassandra/2.1/cassandra/configuration/configGenTokens_c.html) say: "Calculate the tokens for each data center so that the hash range is evenly divided for the nodes in each data center." Why can't each datacenter have exactly the same initial tokens?